src/vumi2/transports/smpp/processors.py

```diff
@@ -87,32 +87,175 @@ def __init__(self, config: dict, sequencer: Sequencer) -> None:
         self.sequencer = sequencer
         self.config = cattrs.structure(config, self.CONFIG_CLASS)
 
-    async def handle_outbound_message(self, message: Message) -> List[PDU]:
+    def _get_msg_length(self, split_msg=False) -> int:
+        # From https://www.twilio.com/docs/glossary/what-sms-character-limit
+        if self.config.data_coding in (
+            DataCodingDefault.SMSC_DEFAULT_ALPHABET,
+            DataCodingDefault.IA5_ASCII,
+        ):
+            if split_msg:
+                return 153
+            else:
+                return 160
+        else:
+            if split_msg:
+                return 134
+            else:
+                return 140
+
+    def _fits_in_one_message(self, content: bytes) -> bool:
+        return len(content) <= self._get_msg_length()
+
+    async def handle_outbound_message(self, message: Message) -> List[SubmitSM]:
         """
         Takes an outbound vumi message, and returns the PDUs necessary to send it.
         """
         # TODO: support USSD over SMPP
         codec = DataCodingCodecs[self.config.data_coding.name]
-        short_message = (message.content or "").encode(codec.value)
+        message_content = (message.content or "").encode(codec.value)
+
+        if (
+            self.config.multipart_handling == MultipartHandling.short_message
+            or self._fits_in_one_message(message_content)
+        ):
+            return [
+                await self.submit_sm_short_message(
+                    message.from_addr, message.to_addr, message_content
+                )
+            ]
+        elif self.config.multipart_handling == MultipartHandling.message_payload:
+            return [
+                await self.submit_sm_long_message(
+                    message.from_addr, message.to_addr, message_content
+                )
+            ]
+        elif self.config.multipart_handling == MultipartHandling.multipart_sar:
+            return await self.submit_csm_sar_message(
+                message.from_addr, message.to_addr, message_content
+            )
+        else:  # multipart_udh:
+            return await self.submit_csm_udh_message(
+                message.from_addr, message.to_addr, message_content
+            )
+
+    async def submit_sm_short_message(
+        self, from_addr: str, to_addr: str, short_message: bytes
+    ) -> SubmitSM:
+        return SubmitSM(
+            seqNum=await self.sequencer.get_next_sequence_number(),
+            service_type=self.config.service_type,
+            source_addr_ton=self.config.source_addr_ton,
+            source_addr_npi=self.config.source_addr_npi,
+            source_addr=from_addr,
+            dest_addr_ton=self.config.dest_addr_ton,
```

Comment:
Is there some commonality we can extract here? Maybe an `async def _submit_sm(self, ...)` method that handles all the statically configured fields?

Reply:
Yeah I think there is, let me take a crack at that